### PR TITLE
HPCC-17644 Initial work on remote key delivery

### PR DESCRIPTION
**Pull request by jakesmith**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [ ] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [ ] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - ready to review, but needs further testing

File: thorlcr/activities/indexread/thindexreadslave.cpp [hunk 1]
```diff
@@ -246,20 +235,24 @@ class CIndexReadSlaveBase : public CSlaveActivity
...
             RemoteFilename rfn;
             part.getFilename(0, rfn);
+            StringBuffer filePath;
+            rfn.getPath(filePath);
+
+            Owned<IDelayedFile> lfile = queryThor().queryFileCache().lookup(*this, logicalFilename, part);
+            Owned<IKeyManager> klManager;
 
-//            if (!rfn.isLocal() && !localKey && !seekGEOffset)
+            //if (!rfn.isLocal() && !localKey && !seekGEOffset)
```
Comment (richardkchapman) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: Is this commented out for testing?

File: thorlcr/graph/thgraphslave.cpp [hunk 2]
```diff
@@ -1775,14 +1775,34 @@ IReplicatedFile *createEnsurePrimaryPartFile(CActivityBase &activity, const char
 class CFileCache;
 class CLazyFileIO : public CInterface, implements IFileIO, implements IDelayedFile
 {
+    class CFileIOWrapper : public CSimpleInterfaceOf<IFileIO>
```
Comment (richardkchapman) [inline thorlcr/graph/thgraphslave.cpp]: Is this change required for this issue or is it something else you spotted needing fixing?

File: common/remote/remoteerr.hpp [hunk 3]
```diff
@@ -66,6 +66,7 @@
 #define RFSERR_VersionMismatch                  8047
 #define RFSERR_SetThrottleFailed                8048
 #define RFSERR_MaxQueueRequests                 8049
+#define RFSERR_KedIndexFailed                   8050
```
Comment (richardkchapman) [inline common/remote/remoteerr.hpp]: Is this supposed to say KeyIndexFailed?

File: common/remote/remoteerr.hpp [hunk 4]
```diff
@@ -86,6 +87,7 @@
 #define RAERR_InvalidUsernamePassword_Text      "Invalid (upper case U) in username/password"
```
Comment (richardkchapman) [inline common/remote/remoteerr.hpp]: Is this still an error??

File: thorlcr/activities/indexread/thindexreadslave.cpp (hunk 1 above)
Comment (jakesmith) [inline thorlcr/activities/indexread/thindexreadslave.cpp]: It's been replaced in a newer commit, where it's also dependent upon a #option.

File: common/remote/remoteerr.hpp (hunk 3 above)
Comment (jakesmith) [inline common/remote/remoteerr.hpp]: yep, I'll change.

File: thorlcr/graph/thgraphslave.cpp (hunk 2 above)
Comment (jakesmith) [inline thorlcr/graph/thgraphslave.cpp]: Should be removed, I will do so.
Though there's a slight caveat/oddity involved here that I'm not sure how to address.
The remote key manager gets the underlying 'real' IFileIO from the lazy io via getRealFileIO().
Which means it can live slightly longer and outside the lazy fileIO cache.

The lazyfileio cache updates the read stats on removal (on IFileIO::close).
Since the remote io doesn't involve any direct IFileIO calls, that's okay as far as these stats are concerned.
Local key stats would be kept via jhtree 
Remote key stats would be tracked remotely, but should probably be captured here too somehow.

File: common/remote/remoteerr.hpp (hunk 4 above)
Comment (jakesmith) [inline common/remote/remoteerr.hpp]: I dunno, no change here, apart from spurious new blank line below it (which I'll remove)

File: dali/dafilesrv/dafilesrv.cpp [hunk 5]
```diff
@@ -691,6 +692,8 @@ int main(int argc,char **argv)
     } perfHook;
     startPerformanceMonitor(10*60*1000, PerfMonStandard, &perfHook);
     writeSentinelFile(sentinelFile);
+
+    setKeyIndexCacheSize(0); // disable IKeyIndex caching
```
Comment (richardkchapman) [inline dali/dafilesrv/dafilesrv.cpp]: Why?

Comment (jakesmith) [inline dali/dafilesrv/dafilesrv.cpp]: Need to discuss.
If enabled, or if not cleared - key writes will not be reflected, i.e. remote key reads will pick up old copies.

File: common/remote/sockfile.cpp [hunk 6]
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
...
+    virtual void setSegmentMonitors(SegMonitorList &segmentMonitors) override { UNIMPLEMENTED; }
+    virtual void deserializeSegmentMonitors(MemoryBuffer &mb) override { UNIMPLEMENTED; }
+    virtual void finishSegmentMonitors() override { }
+    virtual bool lookupSkip(const void *seek, size32_t seekGEOffset, size32_t seeklen) override { UNIMPLEMENTED; }
+    virtual void append(IKeySegmentMonitor *segment) override
+    {
+        segs.append(segment);
+    }
+    virtual unsigned ordinality() const override { UNIMPLEMENTED; }
+    virtual IKeySegmentMonitor *item(unsigned idx) const override { UNIMPLEMENTED; }
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: segs.item(i)?

File: common/remote/sockfile.cpp [hunk 7]
```diff
@@ -3160,6 +3338,15 @@ class CClientStatsTable : public OwningStringSuperHashTableOf<CClientStats>
     }
 };
 
+struct OpenFileInfo
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: minor/future: Code is fine as it is.  Chould possibly have this as a a linked structure, filename a StringAttr, and have a local instance of this which is filled in and then appended to a CArrayOf<> if the open succeeded.  As it is there will be 32 bytes being copied per file instead of 8 when 1 is removed.

Should possibly be a hash table or linked list( instead of an array since there will probably be quite a few entries, and elements get added and removed out of order.  (A std::map which uses a red-black tree may give the best of both worlds.)

File: common/remote/sockfile.cpp [hunk 8]
```diff
@@ -3830,6 +4022,78 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
         return false;
     }
 
+    bool associateKeyManager(int handle, IKeyManager *keyManager)
+    {
+        CriticalBlock block(sect);
+        unsigned clientidx;
+        unsigned handleidx;
+        if (!findHandle(handle,clientidx,handleidx))
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: Out of interest, what is the average number of open files on a dafilesrv?

File: common/remote/sockfile.cpp [hunk 9]
```diff
@@ -3830,6 +4022,78 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+
+    IKeyManager *lookupKM(int handle, const char *keyname)
+    {
+        Owned<IKeyManager> keyManager;
+        CriticalBlock block(sect);
+        unsigned clientidx;
+        unsigned handleidx;
+        if (!findHandle(handle,clientidx,handleidx))
+        {
+            VStringBuffer msg("Failed to find existing key manager for contiuation. key = %s", keyname);
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: typo: conti*n*uation

File: common/remote/sockfile.cpp [hunk 10]
```diff
@@ -3830,6 +4022,78 @@ class CRemoteFileServer : implements IRemoteFileServer, public CInterface
...
+            throw createDafsException(RFSERR_InvalidFileIOHandle, msg.str());
+        }
+        CRemoteClientHandler &client = clients.item(clientidx);
+        IKeyManager *kM = client.openFiles.item(handleidx).keyManager;
+        if (nullptr == kM)
+        {
+            VStringBuffer msg("Failed to find existing key manager for contiuation. key = %s", keyname);
```
Comment (ghalliday) [inline common/remote/sockfile.cpp]: typo: conti*n*uation

File: dali/dafilesrv/dafilesrv.cpp (hunk 5 above)
Comment (ghalliday) [inline dali/dafilesrv/dafilesrv.cpp]: We don't generally overwrite files once written, what is the likely affect of disabling the cache?

**Review by ghalliday — COMMENTED** [review]
@jakesmith see comments from a first review

File: dali/dafilesrv/dafilesrv.cpp (hunk 5 above)
Comment (jakesmith) [inline dali/dafilesrv/dafilesrv.cpp]: I think we need to revisit this separately.
In new cleaned up PR (about to open) I have a slightly better solution I think, where instead of disabling, I keep the cache and flush the specific key on client close.
That at least avoids the problem you do hit otherwise where a updated key (e.g. format) hits errors.

I do wonder if disabling the cache completely might be best in dafilesrv and therefore instead rely on the disk cache.
Unfortunately setKeyIndexCacheSize(0) does not really disable it completely, the jhtree key code is so heavily inter-twined with the node cache that it will still use it.
It is in need of some refactoring as we have discussed before for other reasons too.

File: common/remote/sockfile.cpp [hunk 11]
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
...
+        rowDataBuffer.clear();
+        sizeRemaining = 0;
+        currentSize = 0;
+        currentFpos = 0;
+        currentRow = nullptr;
+        first = true;
+        maxRecsPerRequest = 100;
+        totalGot = 0;
+    }
+    virtual void releaseSegmentMonitors() override { segs.segMonitors.kill(); }
```
Comment (jakesmith) [inline common/remote/sockfile.cpp]: Agree. I think I added SegMonitorList::reset() later and forgot to change this.

File: common/remote/sockfile.cpp [hunk 12]
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
...
+                return true;
+            }
+            else
+            {
+                size32_t maxRecs = maxRecsPerRequest;
+                if (maxRecs && chooseNLimit)
+                {
+                    if (totalGot >= chooseNLimit)
+                        break;
+                    maxRecs = chooseNLimit - totalGot;
```
Comment (jakesmith) [inline common/remote/sockfile.cpp]: agree, will change

File: common/remote/sockfile.cpp [hunk 13]
```diff
@@ -2582,6 +2593,172 @@ void CRemoteFile::copyTo(IFile *dest, size32_t buffersize, ICopyFileProgress *pr
...
+    virtual void setLayoutTranslator(IRecordLayoutTranslator * trans) override { UNIMPLEMENTED; }
+    virtual void setSegmentMonitors(SegMonitorList &segmentMonitors) override { UNIMPLEMENTED; }
+    virtual void deserializeSegmentMonitors(MemoryBuffer &mb) override { UNIMPLEMENTED; }
+    virtual void finishSegmentMonitors() override { }
+    virtual bool lookupSkip(const void *seek, size32_t seekGEOffset, size32_t seeklen) override { UNIMPLEMENTED; }
+    virtual void append(IKeySegmentMonitor *segment) override
+    {
+        segs.append(segment);
+    }
+    virtual unsigned ordinality() const override { UNIMPLEMENTED; }
```
Comment (jakesmith) [inline common/remote/sockfile.cpp]: yes could be, although not foreseen to be used yet. A couple of other seg methods could be implemented too.

File: common/remote/sockfile.cpp (hunk 6 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: yes

File: common/remote/sockfile.cpp (hunk 7 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: Yes, room for improvement here. I thought should be hash table, but didn't want to change at this juncture.

File: common/remote/sockfile.cpp (hunk 8 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: I don't know, but one of the reasons I thought (surprised wasn't / had forgotten) that it was using a hash table for lookup..

File: common/remote/sockfile.cpp (hunk 9 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: code has changed in this area, now uses serialized key cursor.

File: common/remote/sockfile.cpp (hunk 10 above)
Comment (jakesmith) [inline common/remote/sockfile.cpp]: as above

**Comment by jakesmith** [comment]
Closing in favour of cleaner PR.
Review changes will be most recent commit.
